feat(DPE-7464): Add security docs (#87)
canonical/charmed-etcd-operator: docs/explanation/security/cryptography.md

# Cryptography

This document describes the cryptography used by Charmed etcd.

## Resource checksums

> charmed-etcd-operator uses the charmed-etcd-snap, where each revision of the charm pins a revision of the snap to provide reproducible environments.

Reviewer: Suggestion: If we use kebab case here, I suggest us to do it with backticks as a literal. Alternatively, we can use it's proper human-readable name, but then there should be white spaces and capitalisation.

```suggestion
`charmed-etcd-operator` uses the `charmed-etcd-snap`, where each revision of the charm pins a revision of the snap to provide reproducible environments.
```

The snap bundles its workload together with all the essential dependencies and tools needed to support the operator’s lifecycle. For further details, refer to the `snapcraft.yaml` file in the [charmed etcd](https://github.com/canonical/charmed-etcd-snap/blob/3.5/edge/snap/snapcraft.yaml) repository.

Every artifact bundled into a snap is verified against its MD5, SHA256, or SHA512 checksum after download. The installation of certified snap into the rock is ensured by snap primitives that verify their squashfs filesystems images GPG signature. For more information on the snap verification process, refer to the [snapcraft.io documentation](https://snapcraft.io/docs/assertions).

## Sources verification

etcd is built by Canonical from upstream source codes on [Launchpad](https://launchpad.net/etcd).

> The Charmed etcd charm and snap are published and released programmatically using release pipelines implemented via GitHub Actions in their respective repositories.

Reviewer: Question: It looks funny to say "Charmed ... charm". Can we fix it? Is there a difference between `etcd charm and snap` and `charmed etcd charm and snap`? Sometimes there is and it might be important.

Author: No there is no difference in this case.

Reviewer: Can we then avoid repeating the `charm` terminology here?

Author: Yes, I agree. I just updated it.

All repositories in GitHub are set up with branch protection rules, requiring:

* new commits to be merged to main branches via pull request with at least 2 approvals from repository maintainers
* developers to sign the [Canonical Contributor License Agreement (CLA)](https://ubuntu.com/legal/contributors)

## Encryption

Charmed etcd can be used to deploy a secure etcd cluster that provides encryption-in-transit capabilities out of the box for:

* Cluster internal communications (node-to-node)
* External clients connections

> To set up a secure connection Charmed etcd need to be integrated with TLS Certificate Provider charms, e.g. self-signed-certificates operator. Certificate Signing Requests (CSRs) are generated for every unit using the tls_certificates_interface library that uses the cryptography Python library to create X.509 compatible certificates. The CSR is signed by the TLS Certificate Provider, returned to the units, and stored in a Juju secret. The relation also provides the CA certificate, which is then loaded from the Juju secret.

Reviewer: Todo: Fix literals and add links.

```suggestion
To set up a secure connection Charmed etcd need to be integrated with TLS Certificate Provider charms, e.g. `self-signed-certificates` operator. Certificate Signing Requests (CSRs) are generated for every unit using the `tls_certificates_interface` library that uses the `cryptography` Python library to create [X.509 compatible certificates](https://charmhub.io/topics/security-with-x-509-certificates). The CSR is signed by the TLS Certificate Provider, returned to the units, and stored in a Juju secret. The relation also provides the CA certificate, which is then loaded from the [Juju secret](https://documentation.ubuntu.com/juju/3.6/reference/secret/).
```

Encryption at rest is currently not supported, although it can be provided by the substrate (cloud or on-premises).

## Authentication

In Charmed etcd, authentication layers can be enabled for:

1. Admin authentication to etcd, always enabled.
2. etcd cluster authentication, enabled by TLS certificate-based authentication.

> 3. Clients authentication to etcd, enabled by TLS certificate-based authentication.

Reviewer: Suggestion: Authentication enabled by authentication feels funny. Can we adjust the wording here?

```suggestion
1. Admin authentication to etcd -- always enabled.
2. etcd cluster authentication -- enabled by adding a TLS certificate provider.
3. Clients authentication to etcd -- enabled by adding a TLS certificate provider.
```

Please check that the suggested statement is correct.

Author: I would use the word `integrating` instead of `adding`.  I will add the suggestion and change the wording.

### Admin authentication to etcd

Authentication of the admin user to etcd is based on the `bcrypt` go module. See the [etcd official documentation](https://etcd.io/docs/v3.6/learning/design-auth-v3/) for more details.

Credentials are exchanged via [Juju secrets](https://canonical-juju.readthedocs-hosted.com/en/latest/user/howto/manage-secrets/).

### etcd cluster authentication

> Authentication among members of a etcd cluster is based on the TLS certificate-based authentication method. Each unit in the cluster gets issued a TLS certificate signed by the CA of the TLS Certificate Provider charm. The certificate is used to authenticate the unit to other members of the cluster.

Reviewer: chore: Fix articles.

```suggestion
Authentication among members of an etcd cluster is based on the TLS certificate-based authentication method. Each unit in a cluster gets issued a TLS certificate signed by a CA of a TLS Certificate Provider charm. The TLS certificate is then used to authenticate the unit to other members of the cluster.
```

P.S. The word `method` fixes the auth&auth problem here.

# Cryptography

This document describes the cryptography used by Charmed etcd.

## Resource checksums

> Charmed etcd uses pinned versions of the Charmed etcd snap to provide reproducible and secure environments.

Reviewer 2: ```suggestion
charmed-etcd-operator uses the charmed-etcd-snap, where each revision of the charm pins a revision of the snap to provide reproducible environments.
```

> The snap packages its workload along with the necessary dependencies and utilities required for the operators’ lifecycle. For more details, see the snaps content in the `snapcraft.yaml` file for [charmed etcd](https://github.com/canonical/charmed-etcd-snap/blob/3.5/edge/snap/snapcraft.yaml).

Reviewer 2: This should probably be rephrased

Author: I rephrased it. Let me know what you think.

Every artifact bundled into a snap is verified against its MD5, SHA256, or SHA512 checksum after download. The installation of certified snap into the rock is ensured by snap primitives that verify their squashfs filesystems images GPG signature. For more information on the snap verification process, refer to the [snapcraft.io documentation](https://snapcraft.io/docs/assertions).

## Sources verification

etcd is built by Canonical from upstream source codes on [Launchpad](https://launchpad.net/etcd).

The Charmed etcd charm and snap are published and released programmatically using release pipelines implemented via GitHub Actions in their respective repositories.

All repositories in GitHub are set up with branch protection rules, requiring:

* new commits to be merged to main branches via pull request with at least 2 approvals from repository maintainers

> * new commits to be signed (e.g. using GPG keys)

Reviewer 2: I don't think we require this

Author: Removed

* developers to sign the [Canonical Contributor License Agreement (CLA)](https://ubuntu.com/legal/contributors)

## Encryption

Charmed etcd can be used to deploy a secure etcd cluster that provides encryption-in-transit capabilities out of the box for:

> * Cluster internal communications

Reviewer 2: ```suggestion
* Cluster internal communications (node-to-node)
```

* External clients connections

> To set up a secure connection Charmed etcd need to be integrated with TLS Certificate Provider charms, e.g. self-signed-certificates operator. Certificate Signing Requests (CSRs) are generated for every unit using the tls_certificates_interface library that uses the cryptography Python library to create X.509 compatible certificates. The CSR is signed by the TLS Certificate Provider, returned to the units, and stored in Juju secret. The relation also provides the CA certificate, which is loaded into Juju secret.

Reviewer 2: I think?
```suggestion
To set up a secure connection Charmed etcd need to be integrated with TLS Certificate Provider charms, e.g. self-signed-certificates operator. Certificate Signing Requests (CSRs) are generated for every unit using the tls_certificates_interface library that uses the cryptography Python library to create X.509 compatible certificates. The CSR is signed by the TLS Certificate Provider, returned to the units, and stored in a Juju secret. The relation also provides the CA certificate, which is then loaded from the Juju secret.
```

Encryption at rest is currently not supported, although it can be provided by the substrate (cloud or on-premises).

## Authentication

In Charmed etcd, authentication layers can be enabled for:

1. Admin authentication to etcd, always enabled.
2. etcd cluster authentication, enabled by TLS certificate-based authentication.
3. Clients authentication to etcd, enabled by TLS certificate-based authentication.

### Admin authentication to etcd

Authentication of the admin user to etcd is based on the `bcrypt` go module. See the [etcd official documentation](https://etcd.io/docs/v3.5/learning/design-auth-v3/) for more details.

Credentials are exchanged via [Juju secrets](https://canonical-juju.readthedocs-hosted.com/en/latest/user/howto/manage-secrets/).

### etcd cluster authentication

Authentication among members of a etcd cluster is based on the TLS certificate-based authentication method. Each unit in the cluster gets issued a TLS certificate signed by the CA of the TLS Certificate Provider charm. The certificate is used to authenticate the unit to other members of the cluster.

### Clients authentication to etcd

> Authentication of clients to etcd is based on the TLS certificate-based authentication method. Each client provides a TLS certificate they use to authenticate themselves to the etcd cluster. The certificate is stored on a local trusted certificate store, which is used to verify whether the client is allowed to access the etcd cluster.

Reviewer 2: Please add a note about the requirements for leaf certificates.

Author: Added
